File: auth-token-used-instead-of-api-key/python/twilio_credential_audit.py
"""Report whether a Twilio account is still running on its auth token.

Read only. GET requests and nothing else: give this an API Key with read access
rather than the account auth token. If you give it the auth token, it will tell
you so, which is the point.
"""
import argparse
import logging
import os
import sys

import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("twilio_credential_audit")

HOST = "https://api.twilio.com"
BASE = HOST + "/2010-04-01"
MESSAGING = "https://messaging.twilio.com/v1"


def credential_kind(username):
    """Which Twilio credential a basic-auth username implies.

    Every client that talks to the REST API sends basic auth, and the username is
    either an SK API Key SID or the AC account SID. If it is the account SID then
    the password beside it is the account auth token: there is no other password
    that pairs with it. No API field reports this, so the username is the only
    read-only tell there is.
    """
    u = str(username or "").strip().upper()
    if u.startswith("SK"):
        return "api-key"
    if u.startswith("AC"):
        return "auth-token"
    return "unknown"


def verdict(keys, workloads=0, running_as="unknown"):
    """Classify the account's credential posture. Pure, so all four outcomes can
    be tested without a network.

    Order matters: running under the auth token is proof, while a key count is
    inference, and the report should lead with whichever it actually knows.

    Returns (state, detail).
    """
    keys = list(keys or [])
    if running_as == "auth-token":
        return ("auth-token",
                "this run authenticated with the account SID as its basic-auth "
                "username, so the password was the account auth token. That is "
                "proof rather than inference: at least one deployment, this one, "
                "holds the account-wide secret. %d API key(s) exist." % len(keys))

    if not keys:
        return ("no-keys",
                "the account has no API keys, so every service that talks to "
                "Twilio is presenting the auth token: one secret, no per-service "
                "revocation, and the same value that signs your webhooks.")

    if workloads and len(keys) < workloads:
        return ("under-keyed",
                "%d API key(s) for %d separately deployed thing(s): some of them "
                "share a credential, and a shared credential cannot be revoked "
                "for one service without breaking the others."
                % (len(keys), workloads))

    return ("keyed", "%d API key(s) for %d workload(s)." % (len(keys), workloads))


def get(session, url, **params):
    r = session.get(url, params=params, timeout=30)
    if r.status_code in (401, 403):
        raise SystemExit("%d from Twilio: check TWILIO_ACCOUNT_SID and that the "
                         "credential belongs to that account with read access"
                         % r.status_code)
    r.raise_for_status()
    return r.json()


def list_keys(session, account, limit=200):
    """Page Keys.json. next_page_uri is a path, not an absolute URL."""
    url = "%s/Accounts/%s/Keys.json" % (BASE, account)
    params = {"PageSize": 50}
    out = []
    while url and len(out) < limit:
        page = get(session, url, **params)
        out.extend(page.get("keys", []))
        nxt = page.get("next_page_uri")
        url, params = (HOST + nxt) if nxt else None, {}
    return out[:limit]


def count_workloads(session, account):
    """An upper bound on how many separate credentials the account should have.

    Messaging Services and TwiML Applications are a proxy for deployed things,
    not a census of them. It is a rough number and the flag exists to replace it.
    """
    services = get(session, "%s/Services" % MESSAGING, PageSize=50)
    apps = get(session, "%s/Accounts/%s/Applications.json" % (BASE, account),
               PageSize=50)
    return len(services.get("services", [])) + len(apps.get("applications", []))


def main():
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("--workloads", type=int, default=None,
                    help="how many services should have their own key; "
                         "skips the Messaging Service and Application count")
    args = ap.parse_args()

    account = os.environ.get("TWILIO_ACCOUNT_SID")
    key = os.environ.get("TWILIO_API_KEY")
    secret = os.environ.get("TWILIO_API_SECRET")
    if not (account and key and secret):
        log.error("set TWILIO_ACCOUNT_SID, TWILIO_API_KEY and TWILIO_API_SECRET "
                  "(an API Key with read access, not the auth token)")
        return 2

    running_as = credential_kind(key)
    log.info("this run is authenticating as: %s", running_as)

    session = requests.Session()
    session.auth = (key, secret)

    keys = list_keys(session, account)
    for entry in keys:
        log.info("  %s  %s  created %s", entry.get("sid", "?"),
                 entry.get("friendly_name") or "(unnamed)",
                 entry.get("date_created", "?"))

    workloads = args.workloads
    if workloads is None:
        workloads = count_workloads(session, account)

    state, detail = verdict(keys, workloads, running_as)
    if state == "keyed":
        log.info("%-12s %s", state, detail)
        return 0

    log.warning("%-12s %s", state, detail)
    log.warning("  repair: POST %s/Accounts/%s/Keys.json FriendlyName={service-name}, "
                "then store the returned sid and secret as the basic-auth pair",
                BASE, account)
    log.warning("  keep the auth token for X-Twilio-Signature validation and "
                "nowhere else")
    return 1


if __name__ == "__main__":
    sys.exit(main())

File: auth-token-used-instead-of-api-key/node/twilio-credential-audit.mjs
/**
 * Report whether a Twilio account is still running on its auth token.
 *
 * Read only. GET requests and nothing else: give this an API Key with read
 * access rather than the account auth token. If you give it the auth token, it
 * will tell you so, which is the point.
 */
const HOST = 'https://api.twilio.com';
const BASE = `${HOST}/2010-04-01`;
const MESSAGING = 'https://messaging.twilio.com/v1';

/**
 * Which Twilio credential a basic-auth username implies. An SK username is an
 * API Key SID; the AC account SID as a username means the password beside it is
 * the account auth token. No API field reports this, so the username is the only
 * read-only tell there is.
 */
export function credentialKind(username) {
  const u = String(username ?? '').trim().toUpperCase();
  if (u.startsWith('SK')) return 'api-key';
  if (u.startsWith('AC')) return 'auth-token';
  return 'unknown';
}

/**
 * Classify the account's credential posture. Pure, so all four outcomes can be
 * tested without a network. Returns [state, detail].
 */
export function verdict(keys, workloads = 0, runningAs = 'unknown') {
  const all = [...(keys ?? [])];
  if (runningAs === 'auth-token') {
    return ['auth-token',
      'this run authenticated with the account SID as its basic-auth username, ' +
      'so the password was the account auth token. That is proof rather than ' +
      'inference: at least one deployment, this one, holds the account-wide ' +
      `secret. ${all.length} API key(s) exist.`];
  }

  if (all.length === 0) {
    return ['no-keys',
      'the account has no API keys, so every service that talks to Twilio is ' +
      'presenting the auth token: one secret, no per-service revocation, and the ' +
      'same value that signs your webhooks.'];
  }

  if (workloads && all.length < workloads) {
    return ['under-keyed',
      `${all.length} API key(s) for ${workloads} separately deployed thing(s): ` +
      'some of them share a credential, and a shared credential cannot be revoked ' +
      'for one service without breaking the others.'];
  }

  return ['keyed', `${all.length} API key(s) for ${workloads} workload(s).`];
}

function authHeader(key, secret) {
  return `Basic ${Buffer.from(`${key}:${secret}`).toString('base64')}`;
}

async function get(auth, url, params = {}) {
  const u = new URL(url);
  for (const [k, v] of Object.entries(params)) u.searchParams.set(k, v);
  const res = await fetch(u, { headers: { Authorization: auth } });
  if (res.status === 401 || res.status === 403) {
    throw new Error(`${res.status} from Twilio: check TWILIO_ACCOUNT_SID and ` +
                    'that the credential belongs to that account with read access');
  }
  if (!res.ok) throw new Error(`${res.status} from ${u.pathname}`);
  return res.json();
}

export async function listKeys(auth, account, limit = 200) {
  let url = `${BASE}/Accounts/${account}/Keys.json`;
  let params = { PageSize: 50 };
  const out = [];
  while (url && out.length < limit) {
    const page = await get(auth, url, params);
    out.push(...(page.keys ?? []));
    url = page.next_page_uri ? HOST + page.next_page_uri : null;
    params = {};
  }
  return out.slice(0, limit);
}

async function countWorkloads(auth, account) {
  const services = await get(auth, `${MESSAGING}/Services`, { PageSize: 50 });
  const apps = await get(auth, `${BASE}/Accounts/${account}/Applications.json`,
                         { PageSize: 50 });
  return (services.services ?? []).length + (apps.applications ?? []).length;
}

async function main() {
  const account = (process.env.TWILIO_ACCOUNT_SID || "dummy-twilio-account-sid");
  const key = (process.env.TWILIO_API_KEY || "dummy-twilio-api-key");
  const secret = (process.env.TWILIO_API_SECRET || "dummy-twilio-api-secret");
  if (!account || !key || !secret) {
    console.error('set TWILIO_ACCOUNT_SID, TWILIO_API_KEY and TWILIO_API_SECRET ' +
                  '(an API Key with read access, not the auth token)');
    process.exitCode = 2;
    return;
  }

  const runningAs = credentialKind(key);
  console.log(`this run is authenticating as: ${runningAs}`);

  const auth = authHeader(key, secret);
  const keys = await listKeys(auth, account);
  for (const entry of keys) {
    console.log(`  ${entry.sid ?? '?'}  ${entry.friendly_name || '(unnamed)'}  ` +
                `created ${entry.date_created ?? '?'}`);
  }

  const flag = process.argv.indexOf('--workloads');
  const workloads = flag === -1
    ? await countWorkloads(auth, account)
    : Number.parseInt(process.argv[flag + 1], 10);

  const [state, detail] = verdict(keys, workloads, runningAs);
  if (state === 'keyed') {
    console.log(`${state.padEnd(12)} ${detail}`);
    return;
  }

  console.warn(`${state.padEnd(12)} ${detail}`);
  console.warn(`  repair: POST ${BASE}/Accounts/${account}/Keys.json ` +
               'FriendlyName={service-name}, then store the returned sid and ' +
               'secret as the basic-auth pair');
  console.warn('  keep the auth token for X-Twilio-Signature validation and ' +
               'nowhere else');
  process.exitCode = 1;
}

// Only run when invoked directly. The test file imports this module, and without
// the guard main() would run there too, fail on the missing credentials and set a
// non-zero exit code that fails the whole test file even as every test passes.
if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch((err) => { console.error(err.message); process.exitCode = 2; });
}
